streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogRecordDeserializationHelper.java

```
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import org.apache.kafka.common.header.internals.RecordHeader;
+
+/**
+ * Changelog records without any headers are considered old format.
+ * New format changelog records will have a version in their headers.
+ * Version 0: This indicates that the changelog records are under version control.
+ * Version 1: This indicates that the changelog records have consistency information.
+ */
+public class ChangelogRecordDeserializationHelper {
+    private static final byte[] V_0_CHANGELOG_VERSION_HEADER_VALUE = {(byte) 0};
```

`V0` means that the changelog records will have headers from now

---

I think we needed both versions before, when we were thinking about changing the key format of the changelog, but now I think we just need one version. Offhand, it seems like the algorithm should be:

```
versionHeader = record.headers().lastHeader(
                ChangelogRecordDeserializationHelper.CHANGELOG_VERSION_HEADER_KEY);
if (versionHeader == null) {
  return
} else {
  switch (versionHeader.value()[0]) {
    case 0:
      vectorHeader = record.headers().lastHeader(Position.VECTOR_KEY);
      if (vectorHeader == null) {
        throw new StreamsException("This should not happen. Consistency is enabled but the changelog " +
          "contains records without consistency information.");
        }
      position.merge(Position.deserialize(ByteBuffer.wrap(vectorHeader.value())));
    default:
      // log a warning because the changelog writer produced a record with a newer version than we understand
      // maybe we want to zero out the position, since we no longer know what position we're at
      // we probably don't want to throw an exception, since we are presumably in the middle of a rolling upgrade
      return;
  }
}
```

In other words, I don't think the consumer actually needs to check `consistencyEnabled` at all, and I don't think that we currently need the `CHANGELOG_VERSION_HEADER_RECORD_DEFAULT` version that's here right now.

But we do actually need to have some provision for what to do if we get a version that's larger than the ones we know how to handle.

What do you think?

---

OK about the one version. I actually used two versions to decouple changelog record versioning from consistency. So, the rationale was that `V0` means versioning and `V1` means consistency. But it also works if we have one version.

I wanted the consumer to check `consistencyEnabled` so that we can make it explicit to users that are doing rolling upgrades that consistency works only if all servers have consistency enabled. For instance, if the active has consistency enabled but the standby doesn't yet, I want to throw an exception. This way, we don't have weird consistency guarantees when querying, where before a rebalance consistency worked whereas after the rebalance, it won't since the standby doesn't have it enabled. 

The other option is to ignore the position information at standbys when restoring which will lead to the problem above during querying but won't force users to do upgrades on all servers.